paprika/tleap.py

```diff
@@ -0,0 +1,574 @@
+        self.template_lines = None
+        self.loadpdb_file = None
+        self.pbc_type = 'cubic'
+        self.buffer_target = '12.0A'
+        self.water_box = 'TIP3PBOX'
+        self.neutralize = True
+        self.counter_cation='Na+'
+        self.counter_anion='Cl-'
+        self.add_ions = None
+        self.output_path='./'
+        self.output_prefix='build'
+
+        ### Advanced Settings: Defaults
+        self.unit = 'model'
+        self.buffer_value = 1.0
+        self.target_waters = 1000
+        self.exponent = 1
+        self.cyc_since_last_exp_change = 0
+        self.max_cycles = 50
+        self.waters_to_remove = None
+        self.add_ion_residues = None
+        self.buffer_val_history = [0]
+        self.wat_added_history = [0]
+
+    def build(self):
+        """
+        Build the tleap.System
+
+        """
+
+        # Check input
+        if self.template_file and self.template_lines:
+            raise Exception('template_file and template_lines cannot both be specified')
+        elif self.template_file:
+            with open(self.template_file, 'r') as f:
+                self.template_lines = f.read().splitlines()
+        elif self.template_lines:
+            for i,line in enumerate(self.template_lines):
+                self.template_lines[i] = line.rstrip()
+        else:
+            raise Exception('Either template_file or template_lines needs to be specified')
+
+        ### DAVE WOULD LIKE TO SEE THIS BECOME A GENERAL UTILITY!!!
+        # Make sure path is low maintenance
+        if self.output_path[-1] != '/':
+            self.output_path += '/'
+
+        # Filter out any interfering lines
+        self.filter_template()
+
+        # Either just write/run tleap, or do solvate
+        if self.pbc_type is None:
+            self.write_input()
+            self.run()
+        else:
+            self.solvate()
+
+    def filter_template(self):
+        """
+        Filter out any template_lines that may interfere with solvation.
+
+        """
+
+        filtered_lines = []
+        for line in self.template_lines:
+            # Find loadpdb line, replace pdb file if necessary, set unit name
+            if re.search('loadpdb', line):
+                words = line.rstrip().replace('=', ' ').split()
+                if self.loadpdb_file is None:
+                    self.loadpdb_file = words[2]
+                self.unit = words[0]
+                filtered_lines.append("{} = loadpdb {}".format(self.unit, self.loadpdb_file))
+            # Remove any included solvation and ionization commands if pbc_type is not None
+            elif self.pbc_type is not None:
+                if not re.search(r"^\s*(addions|addions2|addionsrand|desc|quit|solvate|save)", line, re.IGNORECASE):
+                    filtered_lines.append(line)
+            else:
+                filtered_lines.append(line)
+
+        self.template_lines = filtered_lines
+
+
+    def write_input(self, include_saves=True):
```

Comment:
Add variables to docstring. 

- https://github.com/heavenshell/vim-pydocstring

Reply:
Done.